dockie.1
.\" Automatically generated by Pandoc 2.7.3
.\"
.TH "DOCKIE" "1" "" "" ""
.hy
.SH NAME
.PP
Dockie - manage unprivileged chroot environments
.SH SYNOPSIS
.PP
dockie [-dhv] {exec | image | ls | import | pull | rm | run | search}
.SH DESCRIPTION
.PP
Dockie is a wrapper around PRoot to manage unprivileged chroots with a
familiar interface.
.PP
PRoot, and therefore Dockie, are not security features and should not be
used as such.
PRoot should only be used as a \[lq]soft sandbox\[rq] where you can:
.IP \[bu] 2
test some other distros features,
.IP \[bu] 2
fake root privileges or
.IP \[bu] 2
build packages from source without polluting your environment
.PP
knowing that whatever gets into your PRoot will be able to get out into
your host OS.
.SS Pulling
.PP
Dockie is able to pull packages from either the LXC image server or from
the Docker Hub.
LXC support is native to Dockie and Docker Hub\[cq]s is done through a
contrib script included with the Dockie releases.
.PP
Dockie stores its data inside DOCKER_PATH, if the environment variable
is set, otherwise inside HOME/.local/var/lib/dockie.
.PP
To pull images the image and version need to be speicifed from the
command line in the form of image:version.
Images that have been either pulled from a remote repository are stored
inside the images directory as a tarball together with a file containing
its formatted metadata.
.PP
As the repositories are provided by third-parties there is no built-in
search mechanism.
The search comnmand will output the URL of the repository where the user
can browse the available images.
.PP
Images are not hashed or identified uniquely in any way.
This means that if an ubuntu:latest image is downloaded and later on
another ubuntu:latest image is downloaded which differs from the
previous (i.e.\ an update has been released), Dockie will overwrite the
previous image and will not keep track of different versions.
If an advanced user would need to keep track of different versions of
the same image, the import subcommand should be used.
.SS Importing
.PP
Dockie also allows importing local images as tarballs.
These tarballs need to only contain the root file system as Dockie will
take care of the metadata.
The only accepted format is .tar.
The name of the file will be used as the name of the image so it is
recommended not to use files such as rootfs.tar without renaming them
first.
.SS Running
.PP
The run command is effectively a combination of three steps:
.IP \[bu] 2
pull: if the image is not available locally and is available on the
remote repository Dockie pulls it from the repository.
.IP \[bu] 2
bootstrap: the tarball is extracted into a newly created directory
inside the guests directory.
/etc/profile and /etc/resolv.conf are created at this point.
Note that Dockie will always replace the existing resolv.conf file on
bootstrap but later changes to the file will be preserved.
.IP \[bu] 2
exec: if a command is passed, Dockie will exec the passed command inside
the guest.
.PP
This means that there are two ways to initialise Dockie guests.
The first one is with an initial pull and the run, which will use the
existing downloaded image.
Guests can also be downloaded implictly, i.e.\ if a guest wants to be
initialised with and image that is not inside the images directory,
Dockie will automatically pull that image and initialise the guest.
.PP
Run will always create a new guest and assign it a unique ID.
It extracts the selected image if it is available locally, otherwise it
pulls it from the configured remote repository.
If a command is specified it also runs the command.
.PP
Guests are inside the guests directory and include the rootfs directory
as well as a formatted metadata file.
.SS Executing
.PP
For guests that have already been creating, the exec subcommand will
execute a command inside the guest, usually /bin/sh or /bin/bash, but
any command can be used.
.PP
Each guest has an unique ID assigned to it, which can be checked with
the ls command.
To execute commands on a guest this ID needs to be used.
.PP
When executing PRoot, if a username has been passed as a command-line
argument it looks for the ID in the etc/password file to log in as that
specified user.
With that username it also sets the HOME environment variable.
On exec, the DISPLAY, TERM, BASH_ENV and ENV are passed to the guest.
Both DISPLAY and TERM have the same value as the guest while BASH_ENV
and ENV are set to /etc/profile to force the guest to read the profile
and therefore setting PS1 according to what Dockie specified during the
bootstrap process.
.PP
To avoid removing guests that are being used, when using exec, Dockie
creates a lock file inside the guest directory (but outside of the root)
so that rm can warn against removing locked guests.
If the file is prevented from being deleted (e.g.\ using chattr or
making the parent directory read-only), this will efectively prevent
that guest from being removed from Dockie.
.SS Options
.TP
.B -d
Debug mode
.TP
.B -h
Print usage
.TP
.B -v
Print version information and quit
.TP
.B --gui
If used with exec, mounts /var/lib/dbus/machine-id, /run/shm, /proc and
/dev from the host
.TP
.B --install
If used with exec, this option is useful to safely create and install
packages into the guest rootfs.
It mounts the following files/directories from the host: /etc/host.conf,
/etc/hosts, /etc/nsswitch.conf, /etc/resolv.conf, /dev/, /sys/, /proc/,
/tmp/, /run/shm, HOME and path.
It is equivalent to proot -S.
.TP
.B --user
If used with exec, specify the username.
.SS Subcommands
.TP
.B exec \f[I]id\f[R] \f[I]command\f[R]
Run a command in a root filesystem identified by an id.
.TP
.B image ls
List images
.TP
.B image rm
Remove one or more images
.TP
.B ls
List guest filesystems
.TP
.B import \f[I]file\f[R]
Import the contents from a tarball to create an image
.TP
.B pull \f[I]image\f[R]
Pull an image
.TP
.B rm \f[I]id\f[R]
Remove one or more root filesystems
.TP
.B run \f[I]id\f[R] \f[I]command\f[R]
Run a command in a new root filesystem
.TP
.B search
Print the URL of the remote repository
.SH ENVIRONMENT
.TP
.B DOCKIE_ARCH
Target architecture for pulled images, if unset it defaults to amd64
.TP
.B DOCKIE_PATH
Location for images and guests.
If unset, it defaults to $HOME/.local/var/lib/dockie
.SH SEE ALSO
.PP
proot(1), chroot(1)
.SH AUTHORS
.TP
.B Cristian Ariza
Initial design


dockie.sh
# shellcheck shell=sh

# _init_dir(path, variable name)
_init_dir() {
	eval "$2=$1"
	mkdir -p "$1"
}

# Usage: dockie import file
#
# Import the contents from a tarball to create an image
#
_import() {
	[ "$#" -ne 1 ] && _usage "import"

	[ "$1" = "${1%.tar}" ] && _log_fatal "extension must be .tar"

	image_name="${1##*/}"
	image_name="${image_name%.*}"

	_initdir "$DOCKIE_IMAGES/$image_name" image_path

	cp "$1" "$image_path/rootfs.tar"

	_tag_image "$image_path" "$image_name"
}

# _bootstrap(system, id, name)
_bootstrap() {
	guest_path="$DOCKIE_GUESTS/$2"
	image_path="$DOCKIE_IMAGES/$1"
	_init_dir "$guest_path/rootfs" guest_prefix

	[ ! -d "$image_path" ] && _pull "$1"

	cd "$guest_prefix" || exit 1
	# sometimes tar has errors and this is ok
	tar xf "$image_path/rootfs.tar" || true

	d="$(date '+%Y-%m-%d %H:%M:%S')"

	printf "%-14s%-20s%-21s%s\n" "$2" "$1" "$d" "$3" >"$guest_path/info"

	{
		# shellcheck disable=SC2016
		printf 'PS1="(%s)$PS1"\n' "$id"
		printf 'typeset -r PS1\n' "$id"

		# not sure why PATH is not exported by default
		echo "export PATH PS1"
	} >>"$guest_prefix/etc/profile" 
		# "$guest_prefix/root/.bashrc" \
		# "$guest_prefix/etc/bash.bashrc" >/dev/null

	rm -f "${guest_prefix:?}"/etc/resolv.conf
	cp "${PREFIX-}"/etc/resolv.conf "$guest_prefix/etc/resolv.conf"

	printf %s\\n "$id"
}

# Usage: dockie exec [OPTIONS] ROOTFS COMMAND [ARG...]
#
# Run a command in an existing guest
#
# Options:
# 	-g, --gui      Needed for some graphical programs to run. Will mount
#	                 /dev, /var/lib/dbus/machine-id, /run/shm and /proc
# 	-i, --install  Needed for some package managers to work. Will mount /dev,
#	                 /sys, /proc, /tmp and /run/shm
# 	-u, --user     Specify username
#
_exec() {
	[ "$#" -lt 2 ] && _usage "exec"

	# check if starts with -
	while arg="$1" && shift && [ "${arg#-}" != "$arg" ]; do
		case x"$arg" in
		x--gui | x-g)
			mounts="${mounts-} -b /var/lib/dbus/machine-id -b /run -b /proc -b /dev"
			;;
		x-v) mounts="${mounts-} -b $1" && shift ;;
		x--user | x-u) user="$1" && shift ;;
		x--install | x-i)
			flags="-b /dev -b /sys -b /proc -b /run -i 0 -r"
			;;
		*) _log_fatal "invalid option '$arg'" ;;
		esac
	done

	[ "$#" -eq 0 ] && _usage exec

	guest_path="$DOCKIE_GUESTS/$arg"
	guest_prefix="$guest_path/rootfs"

	[ ! -d "$guest_prefix" ] && _log_fatal "no such guest: $arg"

	passwd="$guest_prefix/etc/passwd"

	if [ -f "$passwd" ]; then
		id="$(awk -F ':' '$1 == "'"${user-root}"'" { print $3 }' "$passwd")"
		guest_home="$(awk -F ':' '$1 == "'"${user-root}"'" { print $6 }' "$passwd")"
	fi

	[ -z "${flags-}" ] && flags="${mounts-} -i ${id:-0} -r"

	flags="-w ${guest_home:-/} $flags"

	# shellcheck disable=SC2015
	# it's not catastrophic if the lock can't be created
	[ ! -f "$guest_path/lock" ] && touch "$guest_path/lock" || true

	trap 'rm -f $guest_path/lock' EXIT

	envs="DISPLAY=${DISPLAY-} TERM=${TERM-} BASH_ENV=/etc/profile ENV=/etc/profile HOME=${guest_home-}"
	[ -n "${PROOT_TMP_DIR-}" ] && envs="PROOT_TMP_DIR=$PROOT_TMP_DIR $envs"

	# shellcheck disable=SC2086
	env -i $envs "$(command -v proot)" $flags "$guest_prefix" "$@"
}

_image_ls() {
	[ "$#" -ne 0 ] && _usage "image C"

	echo "REPOSITORY          CREATED               SIZE"
	cat "$DOCKIE_IMAGES"/*/info 2>/dev/null
}

# Usage: dockie image COMMAND
#
# Manage images
#
# Commands:
#   ls    List images
#   rm    Remove one or more images
#

_image() {
	[ "$#" -eq 0 ] && _image_ls "$@" && exit
	cmd="$1" && shift
	case "$cmd" in
	ls) _image_ls "$@" ;;
	rm)
		[ "$#" -ne 1 ] && _usage "image rm"
		[ ! -d "$DOCKIE_IMAGES/$1" ] && _log_fatal "no such image: $1"
		rm -rf "${DOCKIE_IMAGES:?}/$1"
		;;
	*) _usage "image C" ;;
	esac
}

_images() { _image_ls "$@"; }

# Usage: dockie image rm [OPTIONS] ROOTFS
#
# Remove a guest.
#
_log_fatal() {
	printf '\033[30;31m%s:\033[30;39m %s\n' "${0##*/}" "$*" >&2 && exit 1
}

# Usage: dockie ls
#
# List guests
#
_ls() {
	[ "$#" -ne 0 ] && _usage "ls"

	echo "ROOTFS ID     IMAGE               CREATED              NAME"
	cat "$DOCKIE_GUESTS"/*/info 2>/dev/null
}

_ps() { _ls "$@"; }

# Usage: dockie pull [OPTIONS] NAME
#
# Pull an image or a repository from a registry
#

# _tag_image(path, name)
_tag_image() {
	printf '%-20s%-22s%s\n' "$2 " "$(date '+%Y-%m-%d %H:%M:%S')" \
		"$(du -h "$1/rootfs.tar" | awk '{print $1"B"}')" >"$1/info"
}

# _pull(system)
_pull() {
	[ "$#" -ne 1 ] && _usage "pull"

	! _match "$1" ':' && _log_fatal "need to specify [image]:[version]"

	if _match "$1" '/'; then
		image_path="$DOCKIE_IMAGES/${1%/*}-${1#*/}"
	else image_path="$DOCKIE_IMAGES/$1"; fi

	rm -rf "${image_path:?}"
	mkdir -p "$image_path"

	# shellcheck disable=SC2015
	_get "$image_path" "$1" || _log_fatal "pull failed for $1"

	_tag_image "$image_path" "$1"
}

# Usage: dockie rm [OPTIONS] ROOTFS
#
# Remove a guest.
#
_rm() {
	[ "$#" -ne 1 ] && _usage rm

	guest_path="$DOCKIE_GUESTS/$1"

	[ ! -d "$guest_path" ] && _log_fatal "no such guest '$1'"

	[ -f "$guest_path/lock" ] && _log_fatal "guest is locked or currently" \
		"in use, otherwise do 'rm $guest_path/lock'"

	rm -rf "$guest_path" || _log_fatal "Please remove any remaining files" \
		"manually."
}

_uuid() {
	id="$(cat /proc/sys/kernel/random/uuid)"
	echo "${id##*-}"
}

# Usage: dockie run [OPTIONS] SYSTEM [COMMAND] [ARG...]
#
# Run a command in a new guest.
#
# Options:
#     -n, --name string    Assign a name to the guest'
#

# _run(options..., image_name)
_run() {
	case "x${1-}" in
	x--name | x-n) guest_name="$2" && shift 2 ;;
	"x" | x-*) _usage "run" ;;
	esac

	image_name="$1" && shift
	id="$(_uuid)"

	_bootstrap "$image_name" "$id" "${guest_name:-$image_name}" >/dev/null

	[ "$#" -eq 0 ] && printf %s\\n "$id" && exit 0
	_exec "$id" "$@"
}

# _match(string, substring)
_match() { case x"$1" in *"$2"*) : ;; *) return 1 ;; esac; }

_usage() {
	# grab usages from comments
	awk '
	/^# Usage: dockie '"$1"'/, 0 {
		if ( $1 != "#" ) exit

		gsub(/# ?/,"")
		print ""$0
	}
	' "$0" && exit 1
}

# Usage: dockie [OPTIONS] COMMAND [ARG...]
#
# Options:
# 	-d        Debug mode
# 	-h        Print usage
# 	-v        Print version information and quit
#
# Commands:
#	exec      Run a command in a root filesystem
#	image     List images
#	ls        List root filesystems
#	import    Import the contents from a tarball to create an image
#	pull      Pull an image
#	rm        Remove one or more root filesystems
#	run       Run a command in a new root filesystem
#	search    Search the image server for images
#
# Run 'dockie COMMAND' for more information on a command.
#

set -eu

VERSION="v0.6.2"

HERE="$(
	cd "$(dirname "$0")"
	pwd
)"
export HERE

: "${DOCKIE_PATH:=$HOME/.dockie}"

_init_dir "$DOCKIE_PATH/images" DOCKIE_IMAGES
_init_dir "$DOCKIE_PATH/guests" DOCKIE_GUESTS

# bash does not like ${1-}
[ "$#" -eq 0 ] && set -- ""

case "x$1" in
x-v) printf 'Dockie version %s\n' "$VERSION" && exit 0 ;;
x-d) set -x && shift ;;
x | x-*) _usage "\[O" ;;
esac

type "_$1" >/dev/null 2>&1 || _usage "\[O"

# shellcheck disable=SC2145
"_$@"
